compiler/rustc_middle/src/traits/chalk.rs

```diff
@@ -336,27 +333,14 @@ pub enum ChalkEnvironmentClause<'tcx> {
     TypeFromEnv(Ty<'tcx>),
 }
 
-impl<'tcx> TypeFoldable<'tcx> for &'tcx ty::List<ChalkEnvironmentClause<'tcx>> {
-    fn super_fold_with<F: TypeFolder<'tcx>>(&self, folder: &mut F) -> Self {
-        let v = self.iter().map(|t| t.fold_with(folder)).collect::<SmallVec<[_; 8]>>();
-        folder.tcx().intern_chalk_environment_clause_list(&v)
-    }
-
-    fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.iter().any(|t| t.visit_with(visitor))
-    }
-}
-/// We have to elaborate the environment of a chalk goal *before*
-/// canonicalization. This type wraps the predicate and the elaborated
-/// environment.
+/// A chalk goal is a predicate.
 #[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, HashStable, TypeFoldable)]
-pub struct ChalkEnvironmentAndGoal<'tcx> {
-    pub environment: &'tcx ty::List<ChalkEnvironmentClause<'tcx>>,
+pub struct ChalkGoal<'tcx> {
     pub goal: ty::Predicate<'tcx>,
 }
```

Comment:
This should stay, but `environment` should be `&'tcx ty::List<ty::Predicate<'tcx>>`